### PR TITLE
[FEATURE] WalletServices: add MerklePath method to Bitails service

### LINKED ISSUE #231: [FEATURE] WalletServices: add MerklePath method to Bitails service

### PR DESCRIPTION
**Pull request by ac4ch**
## Description of Changes

Provide a brief description of the changes you've made.

## Linked Issues / Tickets

Reference any related issues or tickets, e.g. "Closes #123".

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have run the linter

### CODE REVIEW

File: pkg/services/internal/bitails/merkle_path.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		Hash string `json:"hash"`
+	} `json:"branches"`
+}
+
+type merkleBranch struct {
+	Hash string `json:"hash"`
+	Pos  string `json:"pos"`
+}
+
+func (bitails *Bitails) MerklePath(ctx context.Context, txID string) (*wdk.MerklePathResult, error) {
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/merkle_path.go]: Receiver names are different

**Review by mgosek-4chain — APPROVED** [review]

File: pkg/services/internal/bitails/merkle_path.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	Index  int      `json:"index"`
+	TxOrId string   `json:"txOrId"`
+	Target string   `json:"target"`
+	Nodes  []string `json:"nodes"`
+}
+
+type branchProofResponse struct {
+	BlockHash  string `json:"blockhash"`
+	MerkleRoot string `json:"merkleRoot"`
+	Branches   []struct {
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/merkle_path.go]: what's the point of the anonymous struct? while a few lines below there is exact struct definition?

I would use that one.

File: pkg/services/internal/bitails/merkle_path.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	blockHeight, err := to.UInt32(txInfo.BlockHeight)
+	if err != nil {
+		return nil, fmt.Errorf("bitails: invalid block height %d for transaction %s: %w", txInfo.BlockHeight, txID, err)
+	}
+	header.Height = blockHeight
+
+	branches := make([]merkleBranch, len(proof.Branches))
+	for i, b := range proof.Branches {
+		branches[i] = merkleBranch{Hash: b.Hash, Pos: b.Pos}
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/merkle_path.go]: Unnecessary mapping - this should already be in the `proof` struct.

File: pkg/services/internal/bitails/merkle_path.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	MerkleRoot string `json:"merkleRoot"`
+	Branches   []struct {
+		Pos  string `json:"pos"`
+		Hash string `json:"hash"`
+	} `json:"branches"`
+}
+
+type merkleBranch struct {
+	Hash string `json:"hash"`
+	Pos  string `json:"pos"`
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/merkle_path.go]: If this Pos is kinda-enum, I suggest to define & use a string enum

File: pkg/services/services.go [hunk 5]
```diff
@@ -73,8 +73,9 @@ func New(logger *slog.Logger, config defs.WalletServices, opts ...func(*options.
 		getMerklePathServices: servicequeue.NewQueue1(
 			logger,
 			"MerklePath",
-			servicequeue.NewService1(arc.ServiceName, arcService.MerklePath),
+			servicequeue.NewService1(bitails.ServiceName, bitailsService.MerklePath),
```
Comment (chris-4chain) [inline pkg/services/services.go]: ARC service should be at the beginning.

File: pkg/services/internal/whatsonchain/utils.go [hunk 6]
```diff
@@ -53,3 +56,109 @@ func firstNonNilError(errs ...error) error {
 	}
 	return nil
 }
+
+func convertTscProofToMerklePath(txid string, index int, nodes []string, blockHeight uint32) (*transaction.MerklePath, error) {
```
Comment (chris-4chain) [inline pkg/services/internal/whatsonchain/utils.go]: This TSC converter should be implemented once and used wherever necessary (in our case, for now, for Bitails and WoC)

**Review by chris-4chain — CHANGES_REQUESTED** [review]

File: pkg/internal/txutils/proof_for_merkle_path.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"fmt"
+
+	"github.com/bsv-blockchain/go-sdk/chainhash"
+	"github.com/bsv-blockchain/go-sdk/transaction"
+	"github.com/go-softwarelab/common/pkg/to"
+)
+
+// ConvertTscProofToMerklePath converts a TSC proof into a MerklePath structure.
+func ConvertTscProofToMerklePath(txid string, index int, nodes []string, blockHeight uint32) (*transaction.MerklePath, error) {
```
Comment (chris-4chain) [inline pkg/internal/txutils/proof_for_merkle_path.go:12]: Do we have unit tests for this algorithm somewhere?
I mean, including corner & edge cases and for more real-live input-data?

File: go.mod [hunk 8]
```diff
@@ -13,6 +13,7 @@ require (
 	github.com/go-viper/mapstructure/v2 v2.3.0
 	github.com/jarcoal/httpmock v1.4.0
 	github.com/mattn/go-sqlite3 v1.14.28
+	github.com/olekukonko/tablewriter v1.0.7
```
Comment (chris-4chain) [inline go.mod]: Since our examples don't have separate `go.mod` I would try to avoid installing dependencies that are only for the examples.

File: pkg/services/internal/bitails/testabilities/utils.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"encoding/hex"
+	"log"
+
+	"github.com/bsv-blockchain/go-sdk/chainhash"
+)
+
+// MustHashFromHex creates a Hash from hex or panics on failure.
+func MustHashFromHex(hex string) *chainhash.Hash {
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/testabilities/utils.go]: This could do `bittails-given` testability - this way, you would have access to `t testing.TB` so, instead of panic, you could use `require.NotNil`
- In this case the name could be: `HashFromHex`.

File: pkg/services/internal/bitails/testabilities/utils.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	h, err := chainhash.NewHashFromHex(hex)
+	if err != nil {
+		log.Panicf("invalid hex for hash: %v", err)
+	}
+	return h
+}
+
+// FakeHeaderHexWithMerkleRoot returns a fake block header hex string with the given Merkle root in little-endian.
+func FakeHeaderHexWithMerkleRoot(merkleRootHex string) string {
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/testabilities/utils.go]: Similar comment as for `MustHashFromHex`

File: pkg/services/internal/bitails/merkle_path.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+	return &wdk.MerklePathResult{
+		Name:        ServiceName,
+		MerklePath:  merklePath,
+		BlockHeader: header,
+		Notes:       wdk.Notes{{When: to.Ptr(time.Now()), What: "getMerklePathSuccess"}},
+	}, nil
+}
+
+func (b *Bitails) merklePathFromBranchProof(ctx context.Context, txID string) (*wdk.MerklePathResult, error) {
+	url := fmt.Sprintf("%s/tx/%s/proof", b.url, txID)
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/merkle_path.go]: Where is similar algorithm in TS-sdk? I can't find it. 
I mean, the one that uses `"/proof"` without `/tsc` suffix

File: pkg/services/internal/bitails/merkle_path.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+	blockHeight, err := to.UInt32(txInfo.BlockHeight)
+	if err != nil {
+		return nil, fmt.Errorf("bitails: invalid block height %d for transaction %s: %w", txInfo.BlockHeight, txID, err)
+	}
+
+	header, err := b.hashToHeader(ctx, proof.Target)
+	if err != nil {
+		return nil, fmt.Errorf("bitails: failed to fetch block header: %w", err)
+	}
+	header.Height = blockHeight
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/merkle_path.go]: This block of code is repeated in `merklePathFromBranchProof`. Please extract helper function to be more DRY.

File: pkg/services/internal/bitails/utils.go [hunk 13]
```diff
@@ -30,3 +34,47 @@ func buildTxStatusURL(baseURL, txID string) (string, error) {
...
+	offset := uint64(0)
+
+	for i, b := range branches {
+		siblingHash, err := chainhash.NewHashFromHex(b.Hash)
+		if err != nil {
+			return nil, fmt.Errorf("invalid branch hash: %w", err)
+		}
+
+		offset = offset / 2
+		if b.Pos == Left {
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/utils.go]: Where did you find the algorithm that converts this into `MerklePath`?

File: pkg/services/internal/bitails/utils.go [hunk 14]
```diff
@@ -30,3 +34,47 @@ func buildTxStatusURL(baseURL, txID string) (string, error) {
...
 }
+
+func convertBranchProofToMerklePath(txid string, branches []merkleBranch, blockHeight uint32) (*transaction.MerklePath, error) {
+	txidHash, err := chainhash.NewHashFromHex(txid)
+	if err != nil {
+		return nil, fmt.Errorf("invalid txid: %w", err)
+	}
+
+	treeHeight := len(branches) + 1
+	path := make([][]*transaction.PathElement, treeHeight)
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/utils.go]: `path` collides with imported package name

**Review by chris-4chain — CHANGES_REQUESTED** [review]

File: pkg/internal/txutils/proof_for_merkle_path.go (hunk 7 above)
Comment (ac4ch) [reply pkg/internal/txutils/proof_for_merkle_path.go:12]: added the test

File: pkg/internal/txutils/proof_for_merkle_path_test.go [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/txutils"
+	"github.com/stretchr/testify/require"
+)
+
+func TestConvertTscProofToMerklePath(t *testing.T) {
```
Comment (chris-4chain) [inline pkg/internal/txutils/proof_for_merkle_path_test.go:11]: No test case for "duplicateNodeMarker"

**Review by chris-4chain — APPROVED** [review]
